usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -5571,10 +6230,32 @@ CK_RV ep11tok_sign_update(STDLL_TokData_t * tokdata, SESSION * session,
     ep11_private_data_t *ep11_data = tokdata->private_data;
     CK_RV rc;
     SIGN_VERIFY_CONTEXT *ctx = &session->sign_ctx;
+    size_t keyblobsize = 0;
+    CK_BYTE *keyblob;
+    OBJECT *key_obj = NULL;
 
     if (!in_data || !in_data_len)
         return CKR_OK;
 
+    rc = h_opaque_2_blob(tokdata, ctx->key, &keyblob, &keyblobsize, &key_obj,
+                          READ_LOCK);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s h_opaque_2_blob, rc=0x%lx\n", __func__, rc);
+        return rc;
+    }
+
+    rc = ep11tok_pkey_check(tokdata, session, key_obj, &ctx->mech);
```

Comment:
To prevent the (theoretical) possibility that ep11tok_pkey_check() returns something else on the sign update of final call than it returned on the init call, we could add a boolean field in the SIGN_VERIFY_CONTEXT (and the other contexts as well).  Instead of calling ep11tok_pkey_check() in the sign update/final, we would just use that boolean to determine which path to take (PKEY or EP11). That way we ensure that we always take the same path that we took at init. 

As of today, ep11tok_pkey_check will most likely always return the same, but it could be that due to APQNs being disabled or getting offline, or new but older APQNs become online, that CKM_IBM_CPACF_WRAP as of a sudden may not be supported any more. Once we dynamically react on APQNs going away or coming online, this is definitely a possibility. So ep11_data->pkey_wrap_supported could change during run time..... IF this happens just in between an init and an update call this might lead to problems.

Reply:
Done. I added a new boolean field pkey_active to the enc/decr and sign/verify contexts. This field is now checked in any encr/decr and sign/verify operations, except when initialising. At init, the field is set either to true when being in pkey mode, or false when being in ep11 mode.